File: src/main/scala/com/vc/medium/L889.scala
package com.vc.medium


class TreeNode889(var _value: Int) {
  var value: Int = _value
  var left: TreeNode889 = null
  var right: TreeNode889 = null
}

object L889 {
  def constructFromPrePost(pre: Array[Int], post: Array[Int]): TreeNode889 = {
    /**
      *          1
      *       2     3
      *     4  5  6  7
      *
      *   //Pre : 1 2 4 5 3 6 7
      *   //Post: 4 5 2 6 7 3 1
      **/
      def solve(preStart: Int, preEnd: Int, postStart: Int, postEnd: Int): TreeNode889 = {
        if(preStart > preEnd || postStart > postEnd) return null
        val tree = new TreeNode889(pre(preStart))
        if(preStart == preEnd) return tree
        var index = postStart
        var found = false
        while(!found && index < postEnd) {
          if(pre(preStart + 1) == post(index)) found = true
          else index += 1
        }
        if(!found) return tree
        tree.left = solve(preStart + 1, preStart + 1 + index - postStart, postStart, index)
        tree.right = solve(preStart + 1 + index - postStart + 1, preEnd, index + 1, postEnd)
        tree
      }
      solve(0, pre.length - 1, 0, post.length - 1)
  }
}

File: src/main/java/com/vc/medium/L889.java
package com.vc.medium;

import com.vc.hard.TreeNode;

import java.util.HashMap;

class L889 {
    private int[] pre, post;
    private HashMap<Integer, Integer> map;
    public TreeNode constructFromPrePost(int[] pre, int[] post) {
        this.pre = pre;
        this.post = post;
        this.map = new HashMap<Integer, Integer>();
        for(int i = 0; i < pre.length; i++) map.put(pre[i], i);
        return helper(0, pre.length - 1, 0, pre.length - 1);
    }

    private TreeNode helper(int prestart, int preend, int poststart, int postend) {
        if(poststart == postend) return new TreeNode(post[poststart]);

        if(prestart > preend) return null;

        int rootValue = post[postend];
        TreeNode root = new TreeNode(rootValue);

        int divider = post[postend - 1];
        int dividerIndex = map.get(divider);

        int leftCount = dividerIndex - prestart - 1;
        int rightCount = preend - dividerIndex + 1;

        root.left = helper(prestart + 1, prestart + leftCount, poststart, poststart + leftCount - 1);
        root.right = helper(prestart + leftCount + 1, preend, postend - rightCount, postend - 1);

        return root;
    }
}

